README.md
# 关于图形面积计算器的实现
## 1. 大致实现思路
我们利用PyQt5中的图形化库，设计处理程序的可视化界面；然后我们利用信号与槽的关联方法
实现页面的跳转；然后我们为按钮的点击信号设计对应的槽与内部逻辑处理。我们充分关注了对于
输入条件的边界值的测试，例如：当输入不为数字，或者无输入时可以得到对应的提示文件。同时
我们设计了两种单位的选择模式，可以自由切换。然后我们也对于本项目进行了代码单元测试。

## 2. 具体的实现过程：
+ 利用PYQT5的拖拽辅助设计来设计交互页面

![拖拽设计](pic1.png)

然后利用代码转换工具生成代码，并对于图片中的组件位置参数进行调整。

+ 设计跳转页面的信号与槽的关联方法
```python
 #翻页操作
    ui0.pushButton.clicked.connect(partial(MainWindow.show, ))
    ui0.pushButton_1.clicked.connect(partial(SecWindow.show, ))
    ui0.pushButton_2.clicked.connect(partial(ThWindow.show, ))
    ui0.pushButton_3.clicked.connect(partial(FoWindow.show, ))

```
这里定义了菜单到各个页面的翻页操作。

+ 对于菜单类和每一个图形计算框定义一个类
```python
#第四个界面（三角形）的类：
class Ui_Form5(object):
    def setupUi(self, Form):
        Form.setObjectName("Form")
        Form.resize(721, 545)
        self.pushButton = QtWidgets.QPushButton(Form)
        self.pushButton.setGeometry(QtCore.QRect(230, 250, 241, 28))
        self.pushButton.setObjectName("pushButton")

        self.label = QtWidgets.QLabel(Form)
        self.label.setGeometry(QtCore.QRect(280, 30, 181, 101))
        self.label.setObjectName("label")

        self.label_2 = QtWidgets.QLabel(Form)
        self.label_2.setGeometry(QtCore.QRect(50, 200, 241, 31))
        self.label_2.setObjectName("label_2")

        self.label_4 = QtWidgets.QLabel(Form)
        self.label_4.setGeometry(QtCore.QRect(50, 150, 241, 31))
        self.label_4.setObjectName("label_4")

        self.label_5 = QtWidgets.QLabel(Form)
        self.label_5.setGeometry(QtCore.QRect(50, 100, 241, 31))
        self.label_5.setObjectName("label_4")

        self.lineEdit = QtWidgets.QLineEdit(Form)
        self.lineEdit.setGeometry(QtCore.QRect(230, 200, 113, 21))
        self.lineEdit.setObjectName("lineEdit")

        self.lineEdit_2 = QtWidgets.QLineEdit(Form)
        self.lineEdit_2.setGeometry(QtCore.QRect(230, 150, 113, 21))
        self.lineEdit_2.setObjectName("lineEdit")

        self.lineEdit_4 = QtWidgets.QLineEdit(Form)
        self.lineEdit_4.setGeometry(QtCore.QRect(230, 100, 113, 21))
        self.lineEdit_4.setObjectName("lineEdit")

        self.label_3 = QtWidgets.QLabel(Form)
        self.label_3.setGeometry(QtCore.QRect(80, 340, 121, 20))
        self.label_3.setObjectName("label_3")

        self.lineEdit_3 = QtWidgets.QLineEdit(Form)
        self.lineEdit_3.setGeometry(QtCore.QRect(250, 340, 261, 21))
        self.lineEdit_3.setObjectName("lineEdit_3")

        self.radioButton = QtWidgets.QRadioButton(Form)
        self.radioButton.setGeometry(QtCore.QRect(380, 200, 115, 19))
        self.radioButton.setObjectName("radioButton")

        self.radioButton_2 = QtWidgets.QRadioButton(Form)
        self.radioButton_2.setGeometry(QtCore.QRect(510, 200, 115, 19))
        self.radioButton_2.setObjectName("radioButton_2")

        self.retranslateUi(Form)
        QtCore.QMetaObject.connectSlotsByName(Form)

    def retranslateUi(self, Form):
        _translate = QtCore.QCoreApplication.translate
        Form.setWindowTitle(_translate("Form", "Form"))
        self.pushButton.setText(_translate("Form", "计算面积！"))
        self.label.setText(_translate("Form", "三角形面积计算器"))
        self.label_2.setText(_translate("Form", "第三条边（请输入正实数）："))
        self.label_4.setText(_translate("Form", "第二条边（请输入正实数）："))
        self.label_5.setText(_translate("Form", "第一条边（请输入正实数）："))
        self.label_3.setText(_translate("Form", "面积为(平方厘米)："))
        self.radioButton.setText(_translate("Form", "cm"))
        self.radioButton_2.setText(_translate("Form", "英寸"))
```
这里定义了三角形面积计算窗口的类，其中的属性就是窗口内的各个组件，而窗口的方法就是setUpui
和retranslateUi，然后对于每个组件属性赋值，包括位置与名称。

+ 关于数据的处理与计算逻辑：
我们让点击某个Button组件的信号可以实现数据的处理。
```python
#三角型处理函数
def Tri_comvert(ui):
    input_1 = ui.lineEdit.text()
    input_2 = ui.lineEdit_2.text()
    input_4 = ui.lineEdit_4.text()
    # 计算面积
    if input_1!=''and input_2!=''and input_4!=''and is_number(input_1)==1 and is_number(input_2)==1 and is_number(input_4)==1 and float(input_1)>=0 and float(input_2)>=0 and float(input_4)>=0:
        a = float(input_1)
        b = float(input_2)
        c = float(input_4)
        s = (a + b + c) / 2
        if a+b > c and a+c > b and b+c > a:
            result = math.sqrt(s * (s - a) * (s - b) * (s - c))
            if ui.radioButton.isChecked() == True:
               result = float(input_1)
               result = format(1 * result * result * math.pi, '.3f')
            if ui.radioButton_2.isChecked() == True:
               result = float(input_1)
               result = format(2.54 * 2.54 * result * result * math.pi, '.3f')
        else:
            result='输入不构成三角形，重新输入！'
    else:
        result='请输入完整合理数据！'
    ui.lineEdit_3.setText(str(result))

```
其中包括对于用户错误操作的提示以及处理逻辑。当然，我们尤其注重对于边界值的判定与检验，
保证我们的窗口不会因为用户的违规操作而崩溃。

+ 运行效果：
![运行界面](pic3.png)

## 3.需要配置的python软件包
软件中包的引用如下：
```python
import sys
from PyQt5.QtWidgets import QApplication, QMainWindow
from functools import partial
import PyQt5.QtCore
from PyQt5 import QtCore, QtGui, QtWidgets
import math
```
图形面积计算器的主函数是：sqare_calculate.py

单元测试代码：test.py

井字棋的主函数是：main.py

历史记录在：history.txt


## 4.代码单元测试
利用重载函数将原文件中的函数带入test.py中测试
```python
   #三角形的计算函数
    def test_Tri_comvert1(self):
        self.assertEqual(Tri_convert('e', 1,2,1, 0), '请输入完整合理数据！')
    def test_Tri_comvert2(self):
        self.assertEqual(Tri_convert('',1, 1, 0,1), '请输入完整合理数据！')
    def test_Tri_comvert3(self):
        self.assertEqual(Tri_convert(1,-1, 1,1, 0), '请输入完整合理数据！')
    def test_Tri_comvert4(self):
        self.assertEqual(Tri_convert(1,3, 1, 0,1),'输入不构成三角形，重新输入！' )
    def test_Tri_comvert5(self):
         self.assertEqual(Tri_convert(3,4,5, 1, 0),str(format(6, '.3f')) )
    def test_Tri_comvert6(self):
        self.assertEqual(Tri_convert(3, 4, 5, 0, 1), str(format(38.710, '.3f')))


```

## 5.小组分工
+ 软件的交互界面设计：张劭鹏
+ 软件计算功能逻辑：董佳俊
+ 软件的代码测试：张劭鹏，董佳俊，许冲，卜飞扬，王禹翔
+ README文档：张劭鹏

## 6.说明
本次作业的提交:[github地址](https://github.com/Zhang-shaop/calculating-about-area-of-graph)

组长：董佳俊 

联系邮箱：2640573106@qq.com

## 7.兴趣拓展
我们小组在学习了《pygame游戏开发实例》之后，也完成了井字棋小游戏的设计。
可以实现如下功能：
+ 选择人机对战或者好友对战
+ 可以选择先手
+ 每局结束之后可以记录历史对局步骤，保存至history文档中
+ 良好的用户交互界面

运行效果：
![选择界面](pygame_little go\pic4.png)

![运行界面](pygame_little go\pic5.png)

## 8.致谢
感谢助教和王老师一周的教授与答疑，感谢一起工作与学习的队友，感谢认真负责的队长董佳俊。



pygame_little go/history
[ , , ]
[ , , ]
[ , , ]
The data is below:
I choose to play with AI!
 I choose O
[X,,]
[,O,]
[,,]

[X,X,]
[,O,]
[,O,]

X  win!
[X,,]
[,,]
[,,O]

[X,O,X]
[,,]
[,,O]

[X,O,X]
[,O,]
[,X,O]

reset!



sqare_calculate.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'sqare_calculate.ui'
#
# Created by: PyQt5 UI code generator 5.15.7
#
# WARNING: Any manual changes made to this file will be lost when pyuic5 is
# run again.  Do not edit this file unless you know what you are doing.


from PyQt5 import QtCore, QtGui, QtWidgets
import sys
from PyQt5.QtWidgets import QApplication, QMainWindow
from functools import partial
import PyQt5.QtCore
from PyQt5 import QtCore, QtGui, QtWidgets
import math

#第一个界面（正方形）的类：
class Ui_Form(object):
    """
    定义一个计算正方形面积的界面类
    """
    def setupUi(self, Form):
        """
        创建该界面，并创建以下的界面元素：
        pushButton 确认计算的按钮；
        label 该页面的标题显示；
        label_2 提示输入数据类型的显示；
        label_3 提示数据输出的显示；
        radioButton 选择计算单位为厘米；
        radioButton_2 选择计算单位为英寸；
        lineEdit 输入计算所需数据；
        lineEdit_3 输出计算结果；
        同时设置了该窗口的大小以及每个界面元素的大小尺寸        
        """
        Form.setObjectName("Form")
        Form.resize(721, 545)
        self.pushButton = QtWidgets.QPushButton(Form)
        self.pushButton.setGeometry(QtCore.QRect(230, 250, 241, 28))
        self.pushButton.setObjectName("pushButton")
        self.label = QtWidgets.QLabel(Form)
        self.label.setGeometry(QtCore.QRect(280, 90, 181, 101))
        self.label.setObjectName("label")
        self.label_2 = QtWidgets.QLabel(Form)
        self.label_2.setGeometry(QtCore.QRect(50, 200, 241, 31))
        self.label_2.setObjectName("label_2")
        self.lineEdit = QtWidgets.QLineEdit(Form)
        self.lineEdit.setGeometry(QtCore.QRect(230, 200, 113, 21))
        self.lineEdit.setObjectName("lineEdit")
        self.label_3 = QtWidgets.QLabel(Form)
        self.label_3.setGeometry(QtCore.QRect(80, 340, 121, 20))
        self.label_3.setObjectName("label_3")
        self.lineEdit_3 = QtWidgets.QLineEdit(Form)
        self.lineEdit_3.setGeometry(QtCore.QRect(250, 340, 261, 21))
        self.lineEdit_3.setObjectName("lineEdit_3")
        self.radioButton = QtWidgets.QRadioButton(Form)
        self.radioButton.setGeometry(QtCore.QRect(380, 200, 115, 19))
        self.radioButton.setObjectName("radioButton")
        self.radioButton_2 = QtWidgets.QRadioButton(Form)
        self.radioButton_2.setGeometry(QtCore.QRect(510, 200, 115, 19))
        self.radioButton_2.setObjectName("radioButton_2")

        self.retranslateUi(Form)
        QtCore.QMetaObject.connectSlotsByName(Form)

    def retranslateUi(self, Form):
        """
        该函数为界面元素赋予新的命名显示
        """
        _translate = QtCore.QCoreApplication.translate
        Form.setWindowTitle(_translate("Form", "Form"))
        self.pushButton.setText(_translate("Form", "计算面积！"))
        self.label.setText(_translate("Form", "正方形面积计算器"))
        self.label_2.setText(_translate("Form", "边长（请输入正实数）："))
        self.label_3.setText(_translate("Form", "面积为(平方厘米)："))
        self.radioButton.setText(_translate("Form", "cm"))
        self.radioButton_2.setText(_translate("Form", "英寸"))

 # 第三个界面（圆形）的类：
class Ui_Form4(object):
    """
    定义一个计算正方形面积的窗口类
    """
    def setupUi(self, Form):
        """
        创建该界面，并创建以下的界面元素：
        pushButton 确认计算的按钮；
        label 该页面的标题显示；
        label_2 提示输入数据类型的显示；
        label_3 提示数据输出的显示；
        radioButton 选择计算单位为厘米；
        radioButton_2 选择计算单位为英寸；
        lineEdit 输入计算所需数据；
        lineEdit_3 输出计算结果；
        同时设置了该窗口的大小以及每个界面元素的大小尺寸；       
        """
        Form.setObjectName("Form")
        Form.resize(721, 545)
        self.pushButton = QtWidgets.QPushButton(Form)
        self.pushButton.setGeometry(QtCore.QRect(230, 250, 241, 28))
        self.pushButton.setObjectName("pushButton")
        self.label = QtWidgets.QLabel(Form)
        self.label.setGeometry(QtCore.QRect(280, 90, 181, 101))
        self.label.setObjectName("label")
        self.label_2 = QtWidgets.QLabel(Form)
        self.label_2.setGeometry(QtCore.QRect(50, 200, 241, 31))
        self.label_2.setObjectName("label_2")
        self.lineEdit = QtWidgets.QLineEdit(Form)
        self.lineEdit.setGeometry(QtCore.QRect(230, 200, 113, 21))
        self.lineEdit.setObjectName("lineEdit")
        self.label_3 = QtWidgets.QLabel(Form)
        self.label_3.setGeometry(QtCore.QRect(80, 340, 121, 20))
        self.label_3.setObjectName("label_3")
        self.lineEdit_3 = QtWidgets.QLineEdit(Form)
        self.lineEdit_3.setGeometry(QtCore.QRect(250, 340, 261, 21))
        self.lineEdit_3.setObjectName("lineEdit_3")
        self.radioButton = QtWidgets.QRadioButton(Form)
        self.radioButton.setGeometry(QtCore.QRect(380, 200, 115, 19))
        self.radioButton.setObjectName("radioButton")
        self.radioButton_2 = QtWidgets.QRadioButton(Form)
        self.radioButton_2.setGeometry(QtCore.QRect(510, 200, 115, 19))
        self.radioButton_2.setObjectName("radioButton_2")

        self.retranslateUi(Form)
        QtCore.QMetaObject.connectSlotsByName(Form)

    def retranslateUi(self, Form):
        """
        该函数为界面元素赋予新的命名显示
        """
        _translate = QtCore.QCoreApplication.translate
        Form.setWindowTitle(_translate("Form", "Form"))
        self.pushButton.setText(_translate("Form", "计算面积！"))
        self.label.setText(_translate("Form", "圆形面积计算器"))
        self.label_2.setText(_translate("Form", "半径（请输入正实数）："))
        self.label_3.setText(_translate("Form", "面积为(平方厘米)："))
        self.radioButton.setText(_translate("Form", "cm"))
        self.radioButton_2.setText(_translate("Form", "英寸"))

#菜单页面的类
class Ui_Form2(object):
    """
    定义了一个初始菜单界面的类，用来显示需要选择哪个计算模式
    """
    def setupUi(self, Form):
        """
        创建该界面，并创建以下的界面元素：
        label 提示选择某种计算模式的标签；
        pushButton 选择计算正方形的面积；
        pushButton_1 选择计算长方形的面积；
        pushButton_2 选择计算圆形的面积；
        pushButton_3 选择计算三角形的面积；
        同时设置了该窗口的大小以及每个界面元素的大小尺寸
        """
        Form.setObjectName("Form")
        Form.resize(721, 545)
        self.pushButton = QtWidgets.QPushButton(Form)
        self.pushButton.setGeometry(QtCore.QRect(230, 250, 241, 28))
        self.pushButton.setObjectName("pushButton")

        self.label = QtWidgets.QLabel(Form)
        self.label.setGeometry(QtCore.QRect(280, 90, 181, 101))
        self.label.setObjectName("label")

        self.pushButton_1 = QtWidgets.QPushButton(Form)
        self.pushButton_1.setGeometry(QtCore.QRect(230, 300, 241, 28))
        self.pushButton_1.setObjectName("pushButton_1")

        self.pushButton_2 = QtWidgets.QPushButton(Form)
        self.pushButton_2.setGeometry(QtCore.QRect(500, 250, 241, 28))
        self.pushButton_2.setObjectName("pushButton_2")

        self.pushButton_3 = QtWidgets.QPushButton(Form)
        self.pushButton_3.setGeometry(QtCore.QRect(500, 300, 241, 28))
        self.pushButton_3.setObjectName("pushButton_3")

        self.retranslateUi(Form)
        QtCore.QMetaObject.connectSlotsByName(Form)

    def retranslateUi(self, Form):
            """
            该函数为界面元素赋予新的命名显示
            """
            _translate = QtCore.QCoreApplication.translate
            Form.setWindowTitle(_translate("Form", "Form"))
            self.pushButton.setText(_translate("Form", "正方形"))
            self.pushButton_1.setText(_translate("Form", "长方形"))
            self.pushButton_2.setText(_translate("Form", "圆形"))
            self.pushButton_3.setText(_translate("Form", "三角形"))
            self.label.setText(_translate("Form", "请选择图形："))    #已修改

#第二个界面（长方形）的类：（form是3）
class Ui_Form3(object):
    """
    定义一个计算长方形面积的窗口类
    """
    def setupUi(self, Form):
        """
        创建该界面，并创建以下的界面元素：
        pushButton 确认计算的按钮；
        label 该页面的标题显示；
        label_2 提示输入长方形长边长度的显示；
        label_3 提示数据输出的显示；
        label_4 提示输入长方形宽边长度的显示；
        radioButton 选择计算单位为厘米；
        radioButton_2 选择计算单位为英寸；
        lineEdit 输入计算所需数据；
        lineEdit_2 输入计算所需数据；
        lineEdit_3 输出计算结果；
        同时设置了该窗口的大小以及每个界面元素的大小尺寸；       
        """
        Form.setObjectName("Form")
        Form.resize(721, 545)
        self.pushButton = QtWidgets.QPushButton(Form)
        self.pushButton.setGeometry(QtCore.QRect(230, 250, 241, 28))
        self.pushButton.setObjectName("pushButton")

        self.label = QtWidgets.QLabel(Form)
        self.label.setGeometry(QtCore.QRect(280, 50, 181, 101))
        self.label.setObjectName("label")

        self.label_2 = QtWidgets.QLabel(Form)
        self.label_2.setGeometry(QtCore.QRect(50, 200, 241, 31))
        self.label_2.setObjectName("label_2")

        self.label_4 = QtWidgets.QLabel(Form)
        self.label_4.setGeometry(QtCore.QRect(50, 150, 241, 31))
        self.label_4.setObjectName("label_4")

        self.lineEdit = QtWidgets.QLineEdit(Form)
        self.lineEdit.setGeometry(QtCore.QRect(230, 200, 113, 21))
        self.lineEdit.setObjectName("lineEdit")

        self.lineEdit_2 = QtWidgets.QLineEdit(Form)
        self.lineEdit_2.setGeometry(QtCore.QRect(230, 150, 113, 21))
        self.lineEdit_2.setObjectName("lineEdit")

        self.label_3 = QtWidgets.QLabel(Form)
        self.label_3.setGeometry(QtCore.QRect(80, 340, 121, 20))
        self.label_3.setObjectName("label_3")

        self.lineEdit_3 = QtWidgets.QLineEdit(Form)
        self.lineEdit_3.setGeometry(QtCore.QRect(250, 340, 261, 21))
        self.lineEdit_3.setObjectName("lineEdit_3")

        self.radioButton = QtWidgets.QRadioButton(Form)
        self.radioButton.setGeometry(QtCore.QRect(380, 200, 115, 19))
        self.radioButton.setObjectName("radioButton")

        self.radioButton_2 = QtWidgets.QRadioButton(Form)
        self.radioButton_2.setGeometry(QtCore.QRect(510, 200, 115, 19))
        self.radioButton_2.setObjectName("radioButton_2")

        self.retranslateUi(Form)
        QtCore.QMetaObject.connectSlotsByName(Form)

    def retranslateUi(self, Form):
        """
        该函数为界面元素赋予新的命名显示
        """
        _translate = QtCore.QCoreApplication.translate
        Form.setWindowTitle(_translate("Form", "Form"))
        self.pushButton.setText(_translate("Form", "计算面积！"))
        self.label.setText(_translate("Form", "长方形面积计算器"))
        self.label_2.setText(_translate("Form", "长为（请输入正实数）："))
        self.label_4.setText(_translate("Form", "宽为（请输入正实数）："))
        self.label_3.setText(_translate("Form", "面积为(平方厘米)："))
        self.radioButton.setText(_translate("Form", "cm"))
        self.radioButton_2.setText(_translate("Form", "英寸"))

#第四个界面（三角形）的类：
class Ui_Form5(object):
    """
    定义一个计算三角形面积的窗口类
    """
    def setupUi(self, Form):
        """
        创建该界面，并创建以下的界面元素：
        pushButton 确认计算的按钮；
        label 该页面的标题显示；
        label_2 提示输入三角形第三条边的显示；
        label_3 提示输入三角形第二条边的显示；
        label_4 提示输入三角形第一条边的显示；
        label_5 提示数据输出的显示；
        radioButton 选择计算单位为厘米；
        radioButton_2 选择计算单位为英寸；
        lineEdit 输入计算所需数据；
        lineEdit_2 输入计算所需数据；
        lineEdit_3 输出计算结果；
        lineEdit_4 输入计算所需数据；
        同时设置了该窗口的大小以及每个界面元素的大小尺寸；       
        """
        Form.setObjectName("Form")
        Form.resize(721, 545)
        self.pushButton = QtWidgets.QPushButton(Form)
        self.pushButton.setGeometry(QtCore.QRect(230, 250, 241, 28))
        self.pushButton.setObjectName("pushButton")

        self.label = QtWidgets.QLabel(Form)
        self.label.setGeometry(QtCore.QRect(280, 30, 181, 101))
        self.label.setObjectName("label")

        self.label_2 = QtWidgets.QLabel(Form)
        self.label_2.setGeometry(QtCore.QRect(50, 200, 241, 31))
        self.label_2.setObjectName("label_2")

        self.label_4 = QtWidgets.QLabel(Form)
        self.label_4.setGeometry(QtCore.QRect(50, 150, 241, 31))
        self.label_4.setObjectName("label_4")

        self.label_5 = QtWidgets.QLabel(Form)
        self.label_5.setGeometry(QtCore.QRect(50, 100, 241, 31))
        self.label_5.setObjectName("label_4")

        self.lineEdit = QtWidgets.QLineEdit(Form)
        self.lineEdit.setGeometry(QtCore.QRect(230, 200, 113, 21))
        self.lineEdit.setObjectName("lineEdit")

        self.lineEdit_2 = QtWidgets.QLineEdit(Form)
        self.lineEdit_2.setGeometry(QtCore.QRect(230, 150, 113, 21))
        self.lineEdit_2.setObjectName("lineEdit")

        self.lineEdit_4 = QtWidgets.QLineEdit(Form)
        self.lineEdit_4.setGeometry(QtCore.QRect(230, 100, 113, 21))
        self.lineEdit_4.setObjectName("lineEdit")

        self.label_3 = QtWidgets.QLabel(Form)
        self.label_3.setGeometry(QtCore.QRect(80, 340, 121, 20))
        self.label_3.setObjectName("label_3")

        self.lineEdit_3 = QtWidgets.QLineEdit(Form)
        self.lineEdit_3.setGeometry(QtCore.QRect(250, 340, 261, 21))
        self.lineEdit_3.setObjectName("lineEdit_3")

        self.radioButton = QtWidgets.QRadioButton(Form)
        self.radioButton.setGeometry(QtCore.QRect(380, 200, 115, 19))
        self.radioButton.setObjectName("radioButton")

        self.radioButton_2 = QtWidgets.QRadioButton(Form)
        self.radioButton_2.setGeometry(QtCore.QRect(510, 200, 115, 19))
        self.radioButton_2.setObjectName("radioButton_2")

        self.retranslateUi(Form)
        QtCore.QMetaObject.connectSlotsByName(Form)

    def retranslateUi(self, Form):
        """
        该函数为界面元素赋予新的命名显示
        """
        _translate = QtCore.QCoreApplication.translate
        Form.setWindowTitle(_translate("Form", "Form"))
        self.pushButton.setText(_translate("Form", "计算面积！"))
        self.label.setText(_translate("Form", "三角形面积计算器"))
        self.label_2.setText(_translate("Form", "第三条边（请输入正实数）："))
        self.label_4.setText(_translate("Form", "第二条边（请输入正实数）："))
        self.label_5.setText(_translate("Form", "第一条边（请输入正实数）："))
        self.label_3.setText(_translate("Form", "面积为(平方厘米)："))
        self.radioButton.setText(_translate("Form", "cm"))
        self.radioButton_2.setText(_translate("Form", "英寸"))

#正方形的处理函数
def Square_convert(ui):
    """
    该函数主要判定参数输入是否合理，如果没有输入规定的数据或没有选择单位,
    则输出提示；如果选择数据均正确,则计算正方形面积,并将结果显示到lineEdit_3中
    """
    input_1 = ui.lineEdit.text()  #这个地方出来问题，不能读入数据了.唯一的可能就在与存在两个窗口出现了问题。
    if input_1==''or is_number(input_1)==0 or float(input_1)<0:
        result="请输入合理数据！"
    elif ui.radioButton.isChecked()==True:
        result = float(input_1)
        result=format(1*result*result,'.3f')
    elif ui.radioButton_2.isChecked()==True:
        result = float(input_1)
        result=format(2.54*2.54*result*result,'.3f')
    else:
        result='请选择单位！'
    ui.lineEdit_3.setText(str(result))

#长方形处理函数
def rectangle_convert(ui):
    """
    该函数主要判定参数输入是否合理，如果没有输入规定的数据或没有选择单位,
    则输出提示；如果选择数据均正确,则计算正方形面积,并将结果显示到lineEdit_3中
    """
    input_1=ui.lineEdit.text()
    input_2=ui.lineEdit_2.text()
    if input_1==''and input_2=='':
        result='请输入长和宽！'
    elif input_1==''and input_2!='':
        result="请输入长！"
    elif input_2==''and input_1!='':
        result="请输入宽！"
    elif is_number(input_1) == 0 or float(input_1) < 0 or is_number(input_2) == 0 or float(input_2) < 0:
            result = "请输入合理数据！"
    else:
        if ui.radioButton.isChecked()==True:
            input_1=float(input_1)
            input_2=float(input_2)
            result=format(1*input_2*input_1,'.3f')
        elif ui.radioButton_2.isChecked()==True:
            input_1 = float(input_1)
            input_2 = float(input_2)
            result = format(2.54*2.54 * input_2 * input_1,'.3f')
        else:
            result = '请选择单位！'
    ui.lineEdit_3.setText(str(result))

#圆形的处理函数
def Circle_convert(ui):
    input_1 = ui.lineEdit.text()  #这个地方出来问题，不能读入数据了.唯一的可能就在与存在两个窗口出现了问题。
    if input_1==''or is_number(input_1)==0 or float(input_1)<0:
        result="请输入合理数据！"
    elif ui.radioButton.isChecked()==True:
        result = float(input_1)
        result=format(1*result*result*math.pi,'.3f')
    elif ui.radioButton_2.isChecked()==True:
        result = float(input_1)
        result=format(2.54*2.54*result*result*math.pi,'.3f')
    else:
        result='请选择单位！'
    ui.lineEdit_3.setText(str(result))

#三角型处理函数
def Tri_comvert(ui):
    input_1 = ui.lineEdit.text()
    input_2 = ui.lineEdit_2.text()
    input_4 = ui.lineEdit_4.text()
    # 计算面积
    if input_1!=''and input_2!=''and input_4!=''and is_number(input_1)==1 and is_number(input_2)==1 and is_number(input_4)==1 and float(input_1)>=0 and float(input_2)>=0 and float(input_4)>=0:
        a = float(input_1)
        b = float(input_2)
        c = float(input_4)
        s = (a + b + c) / 2
        if a+b > c and a+c > b and b+c > a:
            result = math.sqrt(s * (s - a) * (s - b) * (s - c))
            if ui.radioButton.isChecked() == True:
               result = format(1 * result , '.3f')
            if ui.radioButton_2.isChecked() == True:
               result = format(2.54 * 2.54 * result  , '.3f')
        else:
            result='输入不构成三角形，重新输入！'
    else:
        result='请输入完整合理数据！'
    ui.lineEdit_3.setText(str(result))

#辨别是否为正数
def is_number(s):
    try:
        float(s)
        return True
    except ValueError:
        pass
    try:
        import unicodedata
        unicodedata.numeric(s)
        return True
    except(TypeError,ValueError):
        pass
    return False


#主函数
if __name__ == '__main__':
    app = QApplication(sys.argv)

    # 菜单界面
    MenuWindow = QMainWindow()
    ui0 = Ui_Form2()
    ui0.setupUi(MenuWindow)
    MenuWindow.show()

    #第一个界面
    MainWindow = QMainWindow()
    ui = Ui_Form()
    ui.setupUi(MainWindow)
    ui.pushButton.clicked.connect(partial(Square_convert, ui))

    #第二个界面
    SecWindow = QMainWindow()
    ui2 = Ui_Form3()
    ui2.setupUi(SecWindow)
    ui2.pushButton.clicked.connect(partial(rectangle_convert, ui2))

    #第三个界面(圆形）
    ThWindow=QMainWindow()
    ui3 = Ui_Form4()
    ui3.setupUi(ThWindow)
    ui3.pushButton.clicked.connect(partial(Circle_convert, ui3))

    # 第四个界面(三角形）
    FoWindow = QMainWindow()
    ui4 = Ui_Form5()
    ui4.setupUi(FoWindow)
    ui4.pushButton.clicked.connect(partial(Tri_comvert, ui4))

    #翻页操作
    ui0.pushButton.clicked.connect(partial(MainWindow.show, ))
    ui0.pushButton_1.clicked.connect(partial(SecWindow.show, ))
    ui0.pushButton_2.clicked.connect(partial(ThWindow.show, ))
    ui0.pushButton_3.clicked.connect(partial(FoWindow.show, ))

    # 因为这个是退出，所以必须把代码放在这个之前。不然就会显示出无效的提示。
    sys.exit(app.exec_())